File: Solutions/110-BalancedBinaryTree/BalancedBinaryTree.ts
import { TreeNode } from "../../Types";

export function isBalanced(root: TreeNode | null): boolean {
  function dfs(node: TreeNode | null): [boolean, number] {
    if (!node) return [true, 0];

    const [isBalancedLeft, maxHeightLeft] = dfs(node.left);
    const [isBalancedRight, maxHeightRight] = dfs(node.right);
    const maxHeight = Math.max(maxHeightLeft, maxHeightRight);
    const isSubtreeBalanced =
      isBalancedLeft &&
      isBalancedRight &&
      Math.abs(maxHeightLeft - maxHeightRight) <= 1;

    return [isSubtreeBalanced, maxHeight + 1];
  }

  const [isBalanced] = dfs(root);
  return isBalanced;
}

File: Solutions/All/110-BalancedBinaryTree/BalancedBinaryTree.kt
class Solution {
    fun isBalanced(root: TreeNode?): Boolean {
        val (isTreeBalanced) = dfs(root)
        return isTreeBalanced
    }

    fun dfs(node: TreeNode?): Pair<Boolean, Int> {
        if (node == null) return Pair(true, 0)

        val (isBalancedLeft, maxHeightLeft) = dfs(node.left)
        val (isBalancedRight, maxHeightRight) = dfs(node.right)
        val isTreeBalanced =
            isBalancedLeft &&
                isBalancedRight &&
                Math.abs(maxHeightLeft - maxHeightRight) <= 1
        val maxHeight = 1 + Math.max(maxHeightLeft, maxHeightRight)

        return Pair(isTreeBalanced, maxHeight)
    }
}
